### PR TITLE
Add ability to select all text in the buffer

### LINKED ISSUE #1469: Ability to select/copy all the text in the session
# Summary of the new feature/enhancement

Sometimes there are situations when you need to select/copy all the text in the session.

> ConEmu has such functionality (`Menu` → `Edit` → `Copy All`).

# Proposed technical implementation details (optional)

* Keybinding for action `Select All Text`.

May intersect with #715.

### PR DESCRIPTION
**Pull request by carlos-zamora**
Adds the `selectAll` action which can be used to select all text in the buffer (regardless of whether a selection is present).

## References
#3663 - Mark Mode
#4993 - [Scenario] Keyboard selection

## PR Checklist
* [x] Closes #1469
* [x] Tests added/passed

## Detailed Description of the Pull Request / Additional comments
I've made it such that selecting the "entire buffer" really just selects up to the mutable viewport. This seems like a nice QOL improvement since there's generally nothing past that.

When the user selects all, the viewport does not move. This is consistent with CMD behavior and is intended to allow the user to not lose context when selecting everything.

A minor change had to be made to the DxRenderer because this uncovered an underflow issue. Basically, the selection rects were handed to the DxEngine relative to the viewport (which means that some had a negative y-value). At some point, those rects were stored into `size_t`s, resulting in an underflow issue. This caused the renderer to behave strangely when rendering the selection. Generally, these kinds of issues weren't really noticed because selection would always modify a portion of the viewport.

Funny enough, in a way, this satisfies the "mark mode" scenario because the user now has a way to initiate a selection using only the keyboard. Though this isn't ideal, just a fun thing to point out (that's why I'm not closing the mark mode issue).

## Validation Steps Performed
- Verified using DxEngine and AtlasEngine
- select all --> keyboard selection --> start moving the top-left endpoint (and scroll to there)
- select all --> do not scroll automatically

### CODE REVIEW

**Comment by DHowett** [comment]
Clever, with the existing selection being a requirement for selecting all. It might be hard for somebody to discover this, unfortunately.

I somewhat wonder if we should just make the default <kbd>Ctrl+Shift+A</kbd>, always select all even if there is no selection, and then enable <kbd>Ctrl+A</kbd> _in Mark Mode_

**Comment by zadjii** [comment]
> I somewhat wonder if we should just make the default Ctrl+Shift+A, always select all even if there is no selection, and then enable Ctrl+A _in Mark Mode_

Ack. Especially since there's no way to start a selection with the keyboard ATM

**Review by lhecker — APPROVED** [review]

File: src/cascadia/TerminalControl/ControlCore.cpp [hunk 1]
```diff
@@ -392,6 +392,15 @@ namespace winrt::Microsoft::Terminal::Control::implementation
             vkey != VK_SNAPSHOT &&
             keyDown)
         {
+            // check for "select all"
+            if (modifiers.IsCtrlPressed() && vkey == 'A')
```
Comment (lhecker) [inline src/cascadia/TerminalControl/ControlCore.cpp:396]: Making this configurable would be preferable IMO.

**Review by DHowett — CHANGES_REQUESTED** [review]
As discussed!

File: src/cascadia/TerminalControl/ControlCore.cpp (hunk 1 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalControl/ControlCore.cpp:396]: Oh man the debates we've had over this 😝

**Comment by carlos-zamora** [comment]
> Ctrl+Shift+A, always select all even if there is no selection

I'm ok with this, but if that's the case, we should make it a configurable keybinding.

My main reasoning for that is that I really want to separate mark mode stuff and keybindings. For now, I don't want mark mode to be configurable. We'd be introducing a ton of keybindings that are conditional on a selection being present, and that's spooky. If `selectAll` can be an action that ignores if a selection is present, might as well make it configurable.

> and then enable Ctrl+A _in Mark Mode_

works for me, I'll move this logic over to my mark mode branch then.

**Comment by carlos-zamora** [comment]
FYI Most of the new changes were just plumbing the action through, so I'm explicitly _not_ re-requestion a review from @lhecker.

**Review by zadjii-msft — APPROVED** [review]
#shipit

**Review by DHowett — APPROVED** [review]

**Pull request merged**